src/cxy_ctl.erl

```diff
@@ -88,21 +104,117 @@ incr_active_procs(Task_Type) ->
+        _Not_Slow -> not_slow
+    end.
+
+update_slow_times(Task_Type, Task_Fun, Start, Spawn, Done) -> 
+    not_checked = update_times(make_buffer_slow(Task_Type), Task_Type, Task_Fun, Start, Spawn, Done, false),
+    slow.
+
+update_times(Task_Table, Task_Type, Task_Fun, Start, Spawn, Done, Check_Slowness) ->
     Exec_Elapsed  = timer:now_diff(Done, Spawn),
     Spawn_Elapsed = timer:now_diff(Spawn, Start),
     Elapsed = {Task_Fun, Start, Spawn_Elapsed, Exec_Elapsed},
-    ets_buffer:write(Task_Type_Modified, Elapsed).
+    case ets_buffer:write(Task_Table, Elapsed) of
+        {missing_ets_buffer, _} = Error ->
+            Error;
+        _Num_Entries  ->
+            case Check_Slowness of
+                true  -> check_if_slow(Task_Type, Spawn_Elapsed, Exec_Elapsed);
+                false -> not_checked
+            end
+    end.
 
--spec update_spawn_times(atom(), {module(), atom(), list()}, erlang:timestamp(),
-                         erlang:timestamp(), erlang:timestamp()) -> true.
-update_spawn_times(Task_Type, Task_Fun, Start, Spawn, Done) ->
-    update_times(make_buffer_spawn(Task_Type), Task_Fun, Start, Spawn, Done).
+check_if_slow(Task_Type, Spawn_Elapsed, Exec_Elapsed) ->
+    {Spawn_Cma, Exec_Cma, Slow_Factor_As_Percentage} = update_cmas(Task_Type, Spawn_Elapsed, Exec_Elapsed),
+    case       is_slow(Spawn_Elapsed, Spawn_Cma, Slow_Factor_As_Percentage)
+        orelse is_slow(Exec_Elapsed,  Exec_Cma,  Slow_Factor_As_Percentage) of
 
--spec update_inline_times(atom(), {module(), atom(), list()}, erlang:timestamp(),
-                          erlang:timestamp(), erlang:timestamp()) -> true.
-update_inline_times(Task_Type, Task_Fun, Start, Spawn, Done) ->
-    update_times(make_buffer_inline(Task_Type), Task_Fun, Start, Spawn, Done).
+        true  -> is_slow;
+        false -> not_slow
+    end.
+
+%% Slow_Factor is a percentage, so 300 would be 3x the moving average.
+is_slow(_,                    0,           _) -> false;
+is_slow(Sample_Time, Moving_Avg, Slow_Factor_As_Percentage) ->
+    round((Sample_Time / Moving_Avg) * 100) >= Slow_Factor_As_Percentage.
+
+update_cmas(Task_Type, Spawn_Elapsed, Exec_Elapsed) ->
+
+    %% Fetch, compute and update the moving averages...
+    Cma_Key = make_cma_key(Task_Type),
+    [Old_Spawn_Cma, Old_Exec_Cma, Num_Samples, Slow_Factor_As_Percentage] = get_cma_factors(Cma_Key),
+    New_Exec_Cma  = cumulative_moving_avg(Old_Exec_Cma,  Exec_Elapsed,  Num_Samples),
+    New_Spawn_Cma = cumulative_moving_avg(Old_Spawn_Cma, Spawn_Elapsed, Num_Samples),
+    true = update_cma_avgs(Cma_Key, New_Spawn_Cma, New_Exec_Cma),
+
+    %% The previous moving averages are returned for comparison.
+    {Old_Spawn_Cma, Old_Exec_Cma, Slow_Factor_As_Percentage}.
+
+cumulative_moving_avg(      0, New_Case, _Num_Samples) -> New_Case;
+cumulative_moving_avg(Old_Avg, New_Case,  Num_Samples) ->
+    round((Num_Samples * Old_Avg + New_Case) / (Num_Samples + 1)).
```

Comment:
your average seems to run forever - isn't this supposed to cover a window in time?